service/topology_coordinator.cc

```diff
@@ -814,6 +814,7 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
     }
 
     future<group0_guard> global_token_metadata_barrier(group0_guard&& guard, std::unordered_set<raft::server_id> exclude_nodes = {}) {
+        auto version = _topo_sm._topology.version;
```

Comment:
> Topology version may be updated, for example, by executing a RESTful
API call to move a tablet

Why though? The only one who suppose to update the topology version is a topology coordinator. The API suppose to only ask the coordinator to move a tablets. If we will allow anyone to change topology version we will be in a world of pain.

Reply:
Yeah, we could follow that principle and submit a topology request instead, but that has its own problems:

In the case of this API, we want to allow moving tablets concurrently. For the purpose of the test, which wants two migrations to happen concurrently, but also in production - if you use the API to manually rebalance the cluster and have lots of tablets, you don't want to do it one by one, waiting for the previous migration to finish before the next starts. If we supported multiple requests, then coordinator could coalesce multiple move requests into one so that they could start concurrently. Or have 1 request but add some move-tablet-specific table with a queue of requests. But system.tablets is already such a queue, so the current move tablet handler just writes to it. I think going through requests would add unnecessary complication. group0 atomicity should be enough to ensure that tablet topology remains in consistent state.

Regardless of what we decide we should do with this, I don't think it should be a blocker for this PR, since the code to update topology version is already in master and this is a bug fix to unflake the CI.

Reply:
@tgrabiec please answer the above

Reply:
> OK. But I'd rather we re-implement the API to go through the coordinator afterwards.

Please create an issue

Reply:
> Not everything is protected by group0, only "write back" is. The topology coordinator was free to assume that it is "single threaded" and in case of fail-over group0 was there to protect from writing outdated state. Now the topology coordinator may no longer assume that. It needs to protect itself from local concurrent access. It is step back from the design.

Operations which span more than one group0 transaction, like streaming, have to be performed by topology coordinator. Although now that we have the sessions mechanism, it is safe to cancel any such operation outside the coordinator. External modification of group0 state is like another topology coordinator taking over. 

Other operations could be serialized using group0 mechanism alone, like we do with safe schema changes.

The RESTful APIs for starting tablet ops are one example of operations which are safe to execute outside coordinator. Another example is schema changes. Dropping or creating a table modifies tablet metadata outside coordinator, although it does not change the topology version. Perhaps it should too, if we wanted to assume that same topology version == same erm. Should these also go through topology coordinator? Then we pay the cost of complexity of requesting schema changes via topology requests, and lose concurrency with topology operations.

Reply:
> Operations which span more than one group0 transaction, like streaming, have to be performed by topology coordinator

Not really.

> Other operations could be serialized using group0 mechanism alone, like we do with safe schema changes.

They could. In fact before settling on the topology coordinator  we debated a solution without the coordinator in which each node will simply try to advance the state machine further and only one will eventually succeed because of group0 protection. But we settled on the topology coordinator  not because it is impossible to do it distributed like schema change, but because we decided that such design will result in a simpler code. So any operation can be done outside of the topology coordinator by making it aware of possible races like you did in this patch, but this goes against the design to keep it simple.  And I do not see the reason to deviate from initial design for this particular REST API. We can debate some time later if schema changes should change the topology version or should it have its own thing (already have in fact - schema digest) or may be the version should not be part of the topology and modified by a lot of stuff, but this should be completely different discussion.

Reply:
> And I do not see the reason to deviate from initial design for this particular REST API.

The reason is also simplicity, but in a different place. Currently, all REST API handler logic can be in a single method if it can modify group0 in place. If we go with the requests approach, then the handling would have to be spread in several places which have to coordinate with each other. You can see how it gets complicated in the ALTER keyspace PR for RF changes with tablets. If we limit the scope of allowed concurrent modification to non-busy tablets, then the only place which needs to take into account this concurrent modification is the barrier code, which may be worth it.

Reply:
> > And I do not see the reason to deviate from initial design for this particular REST API.
> 
> The reason is also simplicity, but in a different place. Currently, all REST API handler logic can be in a single method if it can modify group0 in place. If we go with the requests approach, then the handling would have to be spread in several places which have to coordinate with each other. You can see how it gets complicated in the ALTER keyspace PR for RF changes with tablets. If we limit the scope of allowed concurrent modification to non-busy tablets, then the only place which needs to take into account this concurrent modification is the barrier code, which may be worth it.

It is more important to keep topology coordinator simple. That complexity will creep everyware. Instead of going through the coordinator we will have more and more code that spins around waiting for the coordinator to become idle and do stuff on its behalf. This is exactly what we decided we do not want to do. This will also cause unneeded `concurrent_modification` exceptions on the coordinator and the barrier at the start of each topology coordinator operation will no longer be optional like @kbr-scylla claimed here: https://github.com/scylladb/scylladb/issues/18793#issuecomment-2122681687.   The more I look at it the more I think we need to move it to topology requests ASAP. It is not hard to add a new request. We have a lot of code that does it. Alter table is only complicated because it involves both schema and topology together. This is not the case here.

Reply:
> > OK. But I'd rather we re-implement the API to go through the coordinator afterwards.
> 
> Please create an issue

https://github.com/scylladb/scylladb/issues/18817

Reply:
> > > And I do not see the reason to deviate from initial design for this particular REST API.
> > 
> > 
> > The reason is also simplicity, but in a different place. Currently, all REST API handler logic can be in a single method if it can modify group0 in place. If we go with the requests approach, then the handling would have to be spread in several places which have to coordinate with each other. You can see how it gets complicated in the ALTER keyspace PR for RF changes with tablets. If we limit the scope of allowed concurrent modification to non-busy tablets, then the only place which needs to take into account this concurrent modification is the barrier code, which may be worth it.
> 
> It is more important to keep topology coordinator simple. That complexity will creep everyware. Instead of going through the coordinator we will have more and more code that spins around waiting for the coordinator to become idle and do stuff on its behalf. This is exactly what we decided we do not want to do. This will also cause unneeded `concurrent_modification` exceptions on the coordinator and the barrier at the start of each topology coordinator operation will no longer be optional like @kbr-scylla claimed here: [#18793 (comment)]

I don't get the `concurrent_modification` argument. Submitting a request is also a group0 operation which can cause concurrent_modification.

Reply:
> > 
> > It is more important to keep topology coordinator simple. That complexity will creep everyware. Instead of going through the coordinator we will have more and more code that spins around waiting for the coordinator to become idle and do stuff on its behalf. This is exactly what we decided we do not want to do. This will also cause unneeded `concurrent_modification` exceptions on the coordinator and the barrier at the start of each topology coordinator operation will no longer be optional like @kbr-scylla claimed here: [#18793 (comment)]
> 
> I don't get the `concurrent_modification` argument. Submitting a request is also a group0 operation which can cause concurrent_modification.

If the saga you are running "on the side" consist of one step then you are right. But if it takes multiple steps each one may cause the error.

Reply:
> > > It is more important to keep topology coordinator simple. That complexity will creep everyware. Instead of going through the coordinator we will have more and more code that spins around waiting for the coordinator to become idle and do stuff on its behalf. This is exactly what we decided we do not want to do. This will also cause unneeded `concurrent_modification` exceptions on the coordinator and the barrier at the start of each topology coordinator operation will no longer be optional like @kbr-scylla claimed here: [#18793 (comment)]
> > 
> > 
> > I don't get the `concurrent_modification` argument. Submitting a request is also a group0 operation which can cause concurrent_modification.
> 
> If the saga you are running "on the side" consist of one step then you are right. But if it takes multiple steps each one may cause the error.

We don't want to run sagas on the side. Here we're just starting them, but they're carried over by topology coordinator. So we have a single group0 operation. Provided we keep to this limited use case, question is, is going through requests worth it.

Reply:
> > > > It is more important to keep topology coordinator simple. That complexity will creep everyware. Instead of going through the coordinator we will have more and more code that spins around waiting for the coordinator to become idle and do stuff on its behalf. This is exactly what we decided we do not want to do. This will also cause unneeded `concurrent_modification` exceptions on the coordinator and the barrier at the start of each topology coordinator operation will no longer be optional like @kbr-scylla claimed here: [#18793 (comment)]
> > > 
> > > 
> > > I don't get the `concurrent_modification` argument. Submitting a request is also a group0 operation which can cause concurrent_modification.
> > 
> > 
> > If the saga you are running "on the side" consist of one step then you are right. But if it takes multiple steps each one may cause the error.
> 
> We don't want to run sagas on the side. Here we're just starting them, but they're carried over by topology coordinator. So we have a single group0 operation. Provided we keep to this limited use case, question is, is going through requests worth it.

I think it is because it leaves the control in topology coordinator hand. Here we try to start a sage while the coordinator is not looking. But may be the coordinator has something more important to do? We need to implement an easy to use global request queue and then initiation a task will be as simple as calling a right function.